[4354] Add support for cursor based pagination in table view DSL

frouene opened the pull request:
Bug: https://github.com/eclipse-sirius/sirius-web/issues/4354

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

## Review

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

frouene committed:
[4354] Add support for cursor based pagination in table view DSL

Bug: https://github.com/eclipse-sirius/sirius-web/issues/4354
Signed-off-by: Florian ROUËNÉ <florian.rouene@obeosoft.com>

sbegaudeau commented on packages/emf/backend/sirius-components-emf-tables/pom.xml:
```diff
@@ -0,0 +1,138 @@
+	<parent>
+		<groupId>org.springframework.boot</groupId>
+		<artifactId>spring-boot-starter-parent</artifactId>
+		<version>3.3.3</version>
```
Sidenote: there's a PR upgrading Spring Boot currently being reviewed, one way or another this will have to be updated depending on which PR gets merged first.

sbegaudeau commented on packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/table/ViewTableDescriptionConverter.java:
```diff
@@ -217,18 +221,37 @@ private Function<VariableManager, List<Object>> getColumnSemanticElementsProvide
-            if (elementDescription.getDomainType() == null || elementDescription.getDomainType().isBlank()) {
-                return new PaginatedData(candidates, false, false, candidates.size());
+            PaginatedData paginatedResult;
+            if (elementDescription.getSemanticCandidatesExpression() != null && !elementDescription.getSemanticCandidatesExpression().isBlank()) {
```
Could you convert one of the two Papaya example to use the view DSL because there's something I don't understand here. How could you choose to perform this for the specifier?

Is it possible for us to support a `DomainType` in the view DSL row semantic elements expression?

sbegaudeau replied:
As I said on the ADR for the pagination [back in October](https://github.com/eclipse-sirius/sirius-web/pull/4126#discussion_r1816221431)

<img width="746" alt="Capture d’écran 2025-01-03 à 10 08 28" src="https://github.com/user-attachments/assets/1b4edbcb-7da9-45ac-bb8c-ce17c97391a2" />

I don't see how `RowDescription#domainType` or `RowDescription#preconditionExpression` can still exist today. So unless I missed something, `TableElementDescription` should be removed since it does not make much sense for both `CellDescription` and `RowDescription` to share a common parent (that's an useless abstraction). On top of that both `RowDescription#domainType` and `RowDescription#preconditionExpression` should be deleted too because we will never be able to guess how they should behave instead of the specifier as said in October.

frouene replied:
Ok, I keep them to provide a flawed solution if someone use domainType and semanticCandidatesExpression.
I will removed them

sbegaudeau replied:
I'm not I understand your comment, I've asked for the removal of `RowDescription#domainType` and `RowDescription#preconditionExpression` I have never said anything about removing `RowDescription#semanticCandidatesExpression`.

frouene commented:
I've removed the `TableElementDescription` and changed the variable for `RowDescription`.
I still have to add an example of Papaya on this representation View.

sbegaudeau commented on packages/emf/backend/sirius-components-emf-tables/README.adoc:
```diff
@@ -0,0 +1,10 @@
+= sirius-components-emf-forms
```
tables not forms

sbegaudeau commented on packages/sirius-web/backend/sirius-web-application/pom.xml:
```diff
@@ -94,6 +94,11 @@
+		<dependency>
+			<groupId>org.eclipse.sirius</groupId>
+			<artifactId>sirius-components-emf-tables</artifactId>
+			<version>2025.1.1</version>
```
2025.1.2

sbegaudeau commented on packages/emf/backend/sirius-components-emf-tables/pom.xml:
```diff
@@ -0,0 +1,138 @@
+	</parent>
+	<groupId>org.eclipse.sirius</groupId>
+	<artifactId>sirius-components-emf-tables</artifactId>
+	<version>2025.1.1</version>
```
2025.1.2

sbegaudeau commented on packages/emf/backend/sirius-components-emf-tables/pom.xml:
```diff
@@ -0,0 +1,138 @@
+			<groupId>org.eclipse.sirius</groupId>
+			<artifactId>sirius-components-spring-tests</artifactId>
+			<version>2025.1.1</version>
+			<scope>test</scope>
```
2025.1.2

sbegaudeau commented on packages/sirius-web/backend/sirius-web-tests-data/src/main/resources/sirius-web-scripts/studio.sql:
```diff
@@ -650,6 +648,19 @@ INSERT INTO project_image (
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
```
That's not necessary

sbegaudeau commented on packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/table/ViewTableDescriptionConverter.java:
```diff
@@ -52,6 +52,10 @@ public class ViewTableDescriptionConverter implements IRepresentationDescription
+    private static final String CANDIDATE_VARIABLE = "candidate";
+
+    private static final String PREDICATE_VARIABLE = "predicate";
+
```
dead code?

sbegaudeau commented on packages/view/backend/sirius-components-view-table/src/main/resources/model/table.ecore:
```diff
@@ -5,13 +5,13 @@
         eType="ecore:EDataType ../../../../../sirius-components-view/src/main/resources/model/view.ecore#//InterpretedExpression"/>
     <eStructuralFeatures xsi:type="ecore:EReference" name="columnDescriptions" upperBound="-1"
-        eType="#//ColumnDescription" containment="true" eKeys="#//TableElementDescription/name"/>
+        eType="#//ColumnDescription" containment="true"/>
```
Shouldn't the ekeys still be there but with ColumnDescription#name? Same comment for row and cell description below

sbegaudeau reviewed (approved):
I'll wait for the build and merge it

sbegaudeau committed:
[releng] Bump version to 2025.1.3

Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>
